ext/cuda/bilinearinterpolation.jl

```diff
@@ -0,0 +1,222 @@
+import ..ClimaInterpolations.BilinearInterpolation:
+    Bilinear,
+    set_source_range!,
+    interpolatebilinear!,
+    get_stencil_bilinear1d,
+    get_dims
+
+"""
+    interpolatebilinear!(
+        ftarget::CuArray{FT,N},
+        bilinear::B,
+        fsource::CuArray{FT,N},
+    ) where {FT,N,B}
+
+This is the `CUDA` implementation of the `interpolatebilinear!` function.
+
+Interpolate `fsource`, defined on source grid, onto the target grid.
+The horizontal source and target grids are defined in `bilinear`.
+Here `fsource` is an N-dimensional array where the last two dimensions are assumed to be horizontal dimensions.
+
+For example, `fsource` can be of size `[n1, n2..., nx, ny]`, where `nx` and `ny` are the horizontal dimensions.
+Single horizontal level is also supported.
+The number of horizontal levels should be same for both source and target arrays.
+
+The interpolation is parallelized by distributing the horizontal target points across a `(nxblocks, nyblocks)` sized CUDA grid.
+All vertical levels are placed on the same CUDA block.
+When `nlevels` exceeds `nthreads`, `nzloops` loops are executed by each thread to process all vertical levels.
+Similarly, when `ntargetx` exceeds `nxblocks` or `ntargety` exceeds `nyblocks`, `nxloops`/`nyloops` are executed
+by all threads in a given block per that direction.
+"""
+function interpolatebilinear!(
+    ftarget::CuArray{FT, N},
+    bilinear::B,
+    fsource::CuArray{FT, N},
+) where {FT, N, B}
+    @assert N ≥ 2
+    (leveldimssource..., nxs, nys) = size(fsource)
+    (leveldimstarget..., nxt, nyt) = size(ftarget)
+
+    nsourcex, nsourcey, ntargetx, ntargety = get_dims(bilinear)
+
+    # ensure number of vertical levels are same between fsource and ftarget
+    @assert leveldimssource == leveldimstarget
+    # ensure horizontal dimensions of fsource match with horizontal source grid dimensions 
+    # in bilinear 
+    @assert nxs == nsourcex && nys == nsourcey
+    # ensure horizontal dimensions of ftarget match with horizontal target grid dimensions
+    # in bilinear
+    @assert nxt == ntargetx && nyt == ntargety
+    # max threadgroups per grid is not officially specified by Metal
+    # It is possible this number is unlimited on newer Apple models (>m3)
+    maxxblocks = 1024
+    maxyblocks = 1024
+    # target n loops per dimension per block/threadgroup
+    targetloopsperblockdimx = 4
+    targetloopsperblockdimy = 4
```

Comment:
Can we remove these hard-coded numbers?

Reply:
Max blocks is query is not provided with Metal. `1024` is a conservatively supported block size on this platform!
Targeting multiple loops (about (4)( per has been found to be more efficient

Reply:
If we keep the hardcoded numbers, it would be good to add a comment explaining why they were chosen

Reply:
I have changed this from the current conservative limit of `1024` blocks per dim to the CUDA physical limit of `65,535` per block dim and moved this to a function. I have added a comment stating the I am using multiple loops per block, to improve efficiency for this function.